### PR TITLE
TSDF preCalculation pixNorm

### LINKED ISSUE #2675: test_rgbd failed in Debug build
```
[ RUN      ] KinectFusion.lowDense
/build/precommit-contrib_linux64_no_opt/opencv_contrib/modules/rgbd/test/test_kinfu.cpp:328: Failure
Expected: (cv::norm(kfPose.rvec() - pose.rvec())) < (rvecThreshold), actual: 0.0449916 vs 0.02
[  FAILED  ] KinectFusion.lowDense (11003 ms)
[ RUN      ] KinectFusion.inequal
/build/precommit-contrib_linux64_no_opt/opencv_contrib/modules/rgbd/test/test_kinfu.cpp:328: Failure
Expected: (cv::norm(kfPose.rvec() - pose.rvec())) < (rvecThreshold), actual: 0.0440948 vs 0.02
[  FAILED  ] KinectFusion.inequal (10927 ms)
```

relates #2629

### PR DESCRIPTION
**Pull request by DumDereDum**
resolves #2675

Results of GPU optimization:
* before:
[ RUN      ] Perf_TSDF.integrate
[ PERFSTAT ]    (samples=72   mean=2.51   median=2.36   min=1.62   stddev=0.54 (21.7%))
* after:
[ RUN      ] Perf_TSDF.integrate
[ PERFSTAT ]    (samples=72   mean=2.21   median=1.90   min=1.33   stddev=0.82 (37.3%))

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ ] I agree to contribute to the project under Apache 2 License.
- [ ] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [ ] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/rgbd/src/opencl/tsdf.cl [hunk 1]
```diff
@@ -27,6 +27,16 @@ static inline float tsdfToFloat(TsdfType num)
     return ( (float) num ) / (-128);
 }
 
+__kernel void preCalculationPixNorm (__global float * pixNorms,
+                                     __global float * xx,
+                                     __global float * yy)
+{    
+    int i = get_global_id(0);
+    int j = get_global_id(1);
+    int idx = i*480 + j;
```
Comment (savuor) [inline modules/rgbd/src/opencl/tsdf.cl:36]: why 480?

File: modules/rgbd/src/opencl/tsdf.cl [hunk 2]
```diff
@@ -150,7 +161,9 @@ __kernel void integrate(__global const char * depthptr,
         if(v == 0)
             continue;
 
-        float pixNorm = length(camPixVec);
+        int idx = projected.x * 480 + projected.y;
```
Comment (savuor) [inline modules/rgbd/src/opencl/tsdf.cl:164]: why 480?

File: modules/rgbd/src/opencl/tsdf.cl (hunk 1 above)
Comment (wezeyf) [reply modules/rgbd/src/opencl/tsdf.cl:36]: Don't no

**Review by wezeyf — APPROVED** [review]

**Review by wezeyf — APPROVED** [review]

File: modules/rgbd/src/opencl/tsdf.cl (hunk 1 above)
Comment (DumDereDum) [reply modules/rgbd/src/opencl/tsdf.cl]: done

File: modules/rgbd/src/opencl/tsdf.cl (hunk 2 above)
Comment (DumDereDum) [reply modules/rgbd/src/opencl/tsdf.cl]: done

File: modules/rgbd/src/tsdf.cpp [hunk 3]
```diff
@@ -475,7 +478,10 @@ static cv::Mat preCalculationPixNorm(Depth depth, const Intr& intrinsics)
     {
         for (int j = 0; j < widht; j++)
         {
-            pixNorm.at<float>(i, j) = sqrtf(x[j] * x[j] + y[i] * y[i] + 1.0f);
+            //pixNorm.at<float>(i, j) = sqrtf(x[j] * x[j] + y[i] * y[i] + 1.0f);
+            //std::cout <<  << std::endl;
+            int idx = i + j * height;
```
Comment (savuor) [inline modules/rgbd/src/tsdf.cpp:483]: The same: should be `i*width +j`

File: modules/rgbd/src/tsdf.cpp [hunk 4]
```diff
@@ -1202,6 +1208,47 @@ void TSDFVolumeGPU::reset()
...
+        throw std::runtime_error("Failed to create kernel: " + errorStr);
+
+    AccessFlag af = ACCESS_READ;
+    UMat tmp1 = pixNorm1.getUMat(af);
+    UMat xx = x.getUMat(af);
+    UMat yy = y.getUMat(af);
+
+    kk.args(ocl::KernelArg::PtrReadWrite(tmp1),
+        ocl::KernelArg::PtrReadWrite(xx),
+        ocl::KernelArg::PtrReadWrite(yy),
```
Comment (savuor) [inline modules/rgbd/src/tsdf.cpp:1240]: `xx` and `yy` should be `PtrReadOnly`

File: modules/rgbd/src/tsdf.cpp [hunk 5]
```diff
@@ -1240,7 +1297,8 @@ void TSDFVolumeGPU::integrate(InputArray _depth, float depthFactor,
            cxy.val,
            dfac,
            truncDist,
-           maxWeight);
+           maxWeight,
+           ocl::KernelArg::PtrReadWrite(pixNorms));
```
Comment (savuor) [inline modules/rgbd/src/tsdf.cpp:1301]: `pixNorms` should be read-only

File: modules/rgbd/src/tsdf.cpp [hunk 6]
```diff
@@ -331,7 +331,13 @@ struct IntegrateInvoker : ParallelLoopBody
                     }
 
                     // norm(camPixVec) produces double which is too slow
-                    float pixNorm = sqrt(v_reduce_sum(camPixVec*camPixVec));
+                    // float pixNorm = pixNorms.at<float>();
```
Comment (savuor) [inline modules/rgbd/src/tsdf.cpp:334]: Please remove extra comments

File: modules/rgbd/src/opencl/tsdf.cl [hunk 7]
```diff
@@ -27,6 +27,18 @@ static inline float tsdfToFloat(TsdfType num)
...
 }
 
+__kernel void preCalculationPixNorm (__global float * pixNorms,
+                                     __global float * xx,
+                                     __global float * yy,
+                                     int width)
+{    
+    int i = get_global_id(0);
+    int j = get_global_id(1);
+    //int idx = i*480 + j;
```
Comment (savuor) [inline modules/rgbd/src/opencl/tsdf.cl:37]: Please remove extra comment

File: modules/rgbd/src/opencl/tsdf.cl [hunk 8]
```diff
@@ -27,6 +27,18 @@ static inline float tsdfToFloat(TsdfType num)
     return ( (float) num ) / (-128);
 }
 
+__kernel void preCalculationPixNorm (__global float * pixNorms,
+                                     __global float * xx,
+                                     __global float * yy,
```
Comment (savuor) [inline modules/rgbd/src/opencl/tsdf.cl:32]: `xx` and `yy` should be const

File: modules/rgbd/src/opencl/tsdf.cl [hunk 9]
```diff
@@ -39,7 +51,8 @@ __kernel void integrate(__global const char * depthptr,
                         const float2 cxy,
                         const float dfac,
                         const float truncDist,
-                        const int maxWeight)
+                        const int maxWeight,
+                        __global float * pixNorms)
```
Comment (savuor) [inline modules/rgbd/src/opencl/tsdf.cl:55]: `pixNorms` should be const

File: modules/rgbd/src/tsdf.cpp (hunk 4 above)
Comment (DumDereDum) [reply modules/rgbd/src/tsdf.cpp:1240]: done

File: modules/rgbd/src/tsdf.cpp (hunk 5 above)
Comment (DumDereDum) [reply modules/rgbd/src/tsdf.cpp:1301]: done

File: modules/rgbd/src/tsdf.cpp (hunk 6 above)
Comment (DumDereDum) [reply modules/rgbd/src/tsdf.cpp:334]: done

File: modules/rgbd/src/opencl/tsdf.cl (hunk 7 above)
Comment (DumDereDum) [reply modules/rgbd/src/opencl/tsdf.cl:37]: done

File: modules/rgbd/src/opencl/tsdf.cl (hunk 8 above)
Comment (DumDereDum) [reply modules/rgbd/src/opencl/tsdf.cl:32]: done

File: modules/rgbd/src/opencl/tsdf.cl (hunk 9 above)
Comment (DumDereDum) [reply modules/rgbd/src/opencl/tsdf.cl:55]: done

**Review by savuor — APPROVED** [review]

**Comment by savuor** [comment]
👍

**Pull request merged**